lib/byron/bench/Latency.hs

```diff
@@ -0,0 +1,463 @@
+measureLatency start finish tvar action = do
+    atomically $ writeTVar tvar []
+    replicateM_ sampleNTimes action
+    extractTimings start finish . reverse <$> readTVarIO tvar
+
+-- | Scan through iohk-monitoring logs and extract time differences between
+-- start and end messages.
+extractTimings
+    :: (a -> Bool) -- ^ Predicate for start message
+    -> (a -> Bool) -- ^ Predicate for end message
+    -> [LogObject a] -- ^ Log messages
+    -> [NominalDiffTime]
+extractTimings isStart isFinish msgs = map2 mkDiff filtered
+  where
+    map2 _ [] = []
+    map2 f (a:b:xs) = (f a b:map2 f xs)
+    map2 _ _ = error "start trace without matching finish trace"
+
+    mkDiff (False, start) (True, finish) = diffUTCTime finish start
+    mkDiff (False, _) _ = error "missing finish trace"
+    mkDiff (True, _) _ = error "missing start trace"
+
+    filtered = mapMaybe filterMsg msgs
+    filterMsg logObj = case loContent logObj of
+        LogMessage msg | isStart msg -> Just (False, getTimestamp logObj)
+        LogMessage msg | isFinish msg -> Just (True, getTimestamp logObj)
+        _ -> Nothing
+    getTimestamp = tstamp . loMeta
+
+withLatencyLogging
+    ::(Tracers IO -> TVar [LogObject ApiLog] -> IO a)
+    -> IO a
+withLatencyLogging action = do
+    tvar <- newTVarIO []
+    cfg <- defaultConfigStdout
+    CM.setMinSeverity cfg Debug
+    bracket (setupTrace_ cfg "bench-latency") (shutdown . snd) $ \(_, sb) -> do
+        action (setupTracers tvar) tvar `onException` do
+            fmtLn "Action failed. Here are the captured logs:"
+            readTVarIO tvar >>= mapM_ (effectuate sb) . reverse
+
+setupTracers :: TVar [LogObject ApiLog] -> Tracers IO
+setupTracers tvar = nullTracers
+    { apiServerTracer = trMessage $ contramap snd (traceInTVarIO tvar) }
+
+benchWithServer
+    :: Tracers IO
+    -> (Context Byron -> IO ())
+    -> IO ()
+benchWithServer tracers action = do
+    ctx <- newEmptyMVar
+    let setupContext bp wAddr = do
+            let baseUrl = "http://" <> T.pack (show wAddr) <> "/"
+            let sixtySeconds = 60*1000*1000 -- 60s in microseconds
+            manager <- (baseUrl,) <$> newManager (defaultManagerSettings
+                { managerResponseTimeout =
+                    responseTimeoutMicro sixtySeconds
+                })
+            faucet <- initFaucet
+            putMVar ctx $ Context
+                { _cleanup = pure ()
+                , _manager = manager
+                , _walletPort = Port . fromIntegral $ unsafePortNumber wAddr
+                , _faucet = faucet
+                , _feeEstimator = \_ -> error "feeEstimator not available"
+                , _blockchainParameters = bp
+                , _target = Proxy
+                }
+    race (takeMVar ctx >>= action) (withServer setupContext) >>=
+        either pure (throwIO . ProcessHasExited "integration")
+  where
+    withServer act =
+        withCardanoNode nullTracer $(getTestData) $ \socketPath block0 (bp,vData) ->
```

Comment:
In spite of adding `nullTracer` there are tons of logs from `cardano-node` . Any idea how to switch them off?

Reply:
Yeah `cardano-node` is a separate process, so any tracer passed here will not affect it. This tracer is for logging how the cardano-node process starts and stops.

You could edit the `configuration.yaml` and change the minimum log level. Or you could send the process output to `/dev/null`. It probably still logs to a file anyway.